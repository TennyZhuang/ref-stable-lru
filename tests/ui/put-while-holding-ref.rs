use ref_stable_lru::LruCache;
use std::num::NonZeroUsize;

fn main() {
    let mut cache: LruCache<&'static str, String> = LruCache::new(NonZeroUsize::new(3).unwrap());

    let _out = cache.scope(|mut handle, mut perm| {
        handle.put("a", "b".to_string(), &mut perm);
        handle.put("b", "c".to_string(), &mut perm);

        let x = handle.get(&"a", &perm).unwrap().as_str();
        let y = handle.get(&"b", &perm).unwrap().as_str();
        let z = handle.get(&"c", &perm).unwrap().as_str();

        // Should fail here since `x`, `y` and `z` have already borrowed `handle`.
        handle.put("c", "d".to_string(), &mut perm);

        [x, y, z].join(" ")
    });
}
