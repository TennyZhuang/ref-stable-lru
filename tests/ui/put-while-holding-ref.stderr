error[E0502]: cannot borrow `perm` as mutable because it is also borrowed as immutable
  --> tests/ui/put-while-holding-ref.rs:16:42
   |
11 |         let x = handle.get(&"a", &perm).unwrap().as_str();
   |                                  ----- immutable borrow occurs here
...
16 |         handle.put("c", "d".to_string(), &mut perm);
   |                                          ^^^^^^^^^ mutable borrow occurs here
17 |
18 |         [x, y, z].join(" ")
   |          - immutable borrow later used here
